### PR TITLE
Allow ellipsis default args in "trivial" non-stub functions

### PR DESCRIPTION
**Pull request by Michael0x2a**
Resolves https://github.com/python/mypy/issues/6531

This diff modifies the checker to allow default arguments to be used in function definitions in non-stub files as long as the body of that function is either empty or contains just a `raise NotImplementedError`. (This is after ignoring docstrings.)

For rationale on why this is useful, see the issue linked above.

One note on function bodies containing only a `raise Blah`. I was worried that letting mypy ignore the ellipsis when the function raises *any* Exception would lead to unsafe behavior -- I didn't want mypy to start allowing code that looks like this:

    def halt(self, reason: str = ...) -> NoReturn:
        raise MyCustomError("Fatal error: " + reason, self.line, self.current_context)

Whitelisting only NotImplementedError seemed like a good enough compromise -- it's unlikely users will use the arguments in non-trivial ways when raising that particular exception. It should also be easy enough to relax this restriction in the future if necessary.

### CODE REVIEW

File: mypy/checker.py [hunk 1]
```diff
@@ -1012,7 +1015,29 @@ def is_trivial_body(self, block: Block) -> bool:
             return True
         elif len(body) > 1:
             return False
+
         stmt = body[0]
+
+        # Bodies that just raise "NotImplementedError()" count as trivial.
```
Comment (msullivan) [inline mypy/checker.py]: Add a doc string for this function, maybe by moving some of the inline comments to it?

**Comment by JukkaL** [comment]
Maybe mention this somewhere in the mypy docs?

File: test-data/unit/fixtures/exception.pyi [hunk 2]
```diff
@@ -13,5 +13,10 @@ class unicode: pass
...
 class ellipsis: pass
 
-class BaseException: pass
+# Note: this is a slight simplification. In Python 2, the inheritance hierarchy
+# is actually Exception -> StandardError -> RuntimeError -> ...
+class BaseException:
+    def __init__(self, *args: object) -> None: ...
 class Exception(BaseException): pass
+class RuntimeError(Exception): pass
+class NotImplementedError(RuntimeError): pass
```
Comment (ilevkivskyi) [inline test-data/unit/fixtures/exception.pyi]: No new line at the end of file.

**Comment by Michael0x2a** [comment]
Sorry, returning to this only now. I added docs.

I did find writing them to be a bit difficult though: I wasn't sure exactly where to add the docs or how to best present the new info. I ended up shoving them into the page about stubs for now.

So, I'm a little leery of just landing this now. If it's not too much bother, would somebody mind doing one last sanity check over the docs I added?

File: docs/source/stubs.rst [hunk 3]
```diff
@@ -62,20 +62,81 @@ Stub file syntax
...
+
+The recommended style is to use ellipses to do so, just like in
+stub files. It is also considered stylistically acceptable to
+throw a ``NotImplementedError`` in cases where the user of the
+code may accidentally call functions with no actual logic.
+
+You can also elide default arguments as long as the function body
+also contains no runtime logic: the function body only contains
+a single ellipsis, the pass statement, or a ``raise NotImplementedError()``.
+It is also ok for the function body to contain a docstring.
```
Comment (JelleZijlstra) [inline docs/source/stubs.rst]:
```suggestion
It is also acceptable for the function body to contain a docstring.
```
"ok" is a bit too informal

**Pull request merged**